src/Debug.cpp

```diff
@@ -0,0 +1,174 @@
+ */
+#ifdef CHESSPP_TRUNC_LOGS
+    #define CHESSPP_LOG_FILE_OPEN_MODE std::ios::trunc
+#else
+    #define CHESSPP_LOG_FILE_OPEN_MODE std::ios::app
+#endif
+
+namespace chesspp
+{
+    namespace
+    {
+        /**
+         * \brief
+         * Replaces std::clog, std::cerr, std::cout with file streams.
+         * 
+         * Calling enableRedirection() will permanently redirect std::clog, std::cerr, and std::cout to
+         * individual files.
+         */
+        class LogUtil
+        {
+            class LogUtil_buffer
+            : public std::streambuf
+            {
+                bool timestamp_on_next_text = true;
+                std::ostream &sink;
+                std::vector<char> buffer;
+
+            public:
+                explicit LogUtil_buffer(std::ostream &sink_, std::size_t buff_sz = 256)
+                : sink(sink_) //can't use {}
+                , buffer(buff_sz + 1) //don't use {}
+                {
+                    sink.clear();
+                    char *base = &buffer.front();
+                    setp(base, base + buffer.size()-1); //reserve an extra char for calls to overflow
+                }
+
+            private:
+                //Copying disabled
+                LogUtil_buffer(LogUtil_buffer const &) = delete;
+                LogUtil_buffer &operator=(LogUtil_buffer const &) = delete;
+
+                //Applies timestamps and flushes buffer
+                bool timestamp_and_flush()
+                {
+                    std::stringstream out;
+                    for(char *p = pbase(), *e = pptr(); p != e; ++p)
+                    {
+                        if(*p == '\n') timestamp_on_next_text = true;
+                        else if(timestamp_on_next_text)
+                        {
+                            if(!std::isspace(*p)) out << timestamp();
+                            timestamp_on_next_text = false;
+                        }
+                        out << *p;
+                    }
+
+                    std::ptrdiff_t n = pptr() - pbase();
+                    pbump(-n);
+
+                    sink << out.str();
+                    sink.flush();
+                    return (!sink.fail() && !sink.bad());
+                }
+
+                //Overridden streambuf functions
+                int_type overflow(int_type ch) override
+                {
+                    if(sink && ch != traits_type::eof())
+                    {
+                        assert(std::less_equal<char *>()(pptr(), epptr()));
+                        *pptr() = ch;
+                        pbump(1);
+                        if(timestamp_and_flush())
+                            return ch;
+                    }
+                    return traits_type::eof();
+                }
+                int sync() override
+                {
+                    return timestamp_and_flush() ? 0 : -1;
+                }
+
+                //returns std::string containing current system time.
```

Comment:
Also not part of this PR, but all functions should be documented in Doxygen format.